draft-ietf-jsonpath-base.md

```diff
@@ -1165,16 +1165,16 @@ The following table lists filter expression operators in order of precedence fro
 |  1  | Logical OR | `¦¦`   |
 {: title="Filter expression operator precedence" }
 
-#### Semantics
+##### Semantics
 {: unnumbered}
 
-The `filter-selector` works with arrays and objects exclusively. Its result is a list of *zero*, *one*, *multiple* or *all* of their array elements or member values, respectively.
+The filter //PICKER// works with arrays and objects exclusively. Its result is a list of *zero*, *one*, *multiple* or *all* of their array elements or member values, respectively.
 Applied to other value types, it will select nothing.
 
 A relative path, beginning with `@`, refers to the current array element or member value as the
-filter selector iterates over the array or object.
+filter //PICKER// iterates over the array or object.
```

Comment:
No. This is selector language again.

Reply:
No. Again, *pickers iterate because not all of them need to.

Reply:
I recommend to avoid terms suggesting implementation details (iterating, recursive, ...) altogether.

What about: 'filter //PICKER// visits components of structured values one after another.' ?

Implementers will know very well how to do it.

I like the strict distinction ...

* *pickers *match* values.
* selectors build a *result list* from those.

Reply:
Without constraining implementations, I think we should define the processing model for how selectors and *pickers work together. I think there is a choice between:

1. a selector presents a single value to a *picker, the *picker matching zero or more children of the value, and the selector builds its resultant node list from the matched children (plus some words about concatenating the node lists together when there is more than one *picker), and
2. a selector presents each child of a value in turn to a *picker along with the value's name (if the value is a child of an object) or index (if the value is a child of an array), the *picker matches the value or doesn't match the value, and the selector builds its resultant node list from the matched children (plus some words about concatenating the node lists together when there is more than one *picker).

I _think_ the behaviour is equivalent in either case. The choice simply influences how each *picker is specified (and how each selector which uses *pickers is specified). I personally favour option 2 because it makes the *picker interface narrower and I think that will make the *picker specs shorter and simpler.

Reply:
My thinking is in line with the first option.

Suppose we have `$[1]` and the data is an array with 1000 items.  It doesn't make sense to describe the selector as presenting each item to the *picker `1` to get a boolean result to determine whether to include it.

> the *picker matches the value or doesn't match the value

A *picker can't do this from just the value.  The `1` in my example needs the context of position within its parent.  _Filters_ specifically function just on the value, but the others don't.  Thus is wrong to generalize this way.

Reply:
> My thinking is in line with the first option.
> 
> Suppose we have `$[1]` and the data is an array with 1000 items. It doesn't make sense to describe the selector as presenting each item to the *picker `1` to get a boolean result to determine whether to include it.

Why not? This is a specification and doesn't imply that an implementation would do the same.

> 
> > the *picker matches the value or doesn't match the value
> 
> A *picker can't do this from just the value. The `1` in my example needs the context of position within its parent.

That's why option 2 says that the value's name or index is also provided to the *picker.

> _Filters_ specifically function just on the value, but the others don't. Thus is wrong to generalize this way.

Not necessarily. If the value and its name/index is passed to the *picker, that provides enough context for all *pickers. Some may ignore some of these "parameters", but again we are talking about a spec rather than an implementation.

Reply:
> Why not? This is a specification and doesn't imply that an implementation would do the same.

For the same reason we try not to prescribe implementation: implementors will default to the process described in the spec.  If the spec describes a known bad way to do something, that's bad for implementors.

It just doesn't make sense to have the iteration done at the selector when ONLY ONE *picker actually does any kind of iteration.  Filter is the only picker that iterates over the values.  It's also the only one that depends on the value at all.  All of the others depend on the value's location in relation to its parent.

It's incorrect to describe this behavior holistically because there is no holistic behavior among the *pickers.

You want to define the relationship between selectors and *pickers.  Here you go:

> The child selector selects values identified by the *pickers it contains by concatenating the resulting sets of values from each *picker in the order they appear without removing duplicates.

The *picker identifies which values to select, and the selector "selects" them.  Yeah, the child selector is little more than a container for *pickers.  That's okay.  It doesn't have to do anything.